pkg/controllers/interruption/controller.go

```diff
@@ -55,71 +55,108 @@ type Controller struct {
 
 	unavailableOfferingsCache *cache.UnavailableOfferings
 	metadataClient            *metadata.Client
-	operationClient           *computev1.RegionOperationsClient
+	zoneOperationClient       *computev1.ZoneOperationsClient
 	credential                auth.Credential
 }
 
 func NewController(kubeClient client.Client, recorder events.Recorder,
 	unavailableOfferingsCache *cache.UnavailableOfferings,
 	metadataClient *metadata.Client,
-	operationClient *computev1.RegionOperationsClient,
+	zoneOperationClient *computev1.ZoneOperationsClient,
 	credential auth.Credential) *Controller {
 	return &Controller{
 		kubeClient: kubeClient,
 		recorder:   recorder,
 
 		unavailableOfferingsCache: unavailableOfferingsCache,
 		metadataClient:            metadataClient,
-		operationClient:           operationClient,
+		zoneOperationClient:       zoneOperationClient,
 		credential:                credential,
 	}
 }
 
-func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
-	// Refer to https://cloud.google.com/compute/docs/instances/create-use-spot#detect-preemption
-	// We need to check if the instance is preempted and delete the nodeClaim from the operation event,
-	// In this stage, we only detect preempted operation here.
-	it := c.operationClient.List(ctx, &computepb.ListRegionOperationsRequest{
-		Project: c.credential.ProjectID,
-		Region:  c.credential.Region,
-		Filter:  swag.String(OperationTypePreempted),
-	})
-
-	instanceNames := []string{}
-	for {
-		op, err := it.Next()
-		if err != nil {
-			log.FromContext(ctx).Error(err, "listing operations")
-			break
+// getZonesFromNodes lists all zones by inspecting node labels
+func getZonesFromNodes(ctx context.Context, kubeClient client.Client) ([]string, error) {
+	nodeList := &corev1.NodeList{}
+	if err := kubeClient.List(ctx, nodeList, &client.ListOptions{}); err != nil {
+		return nil, fmt.Errorf("listing nodes: %w", err)
+	}
+
+	zoneSet := make(map[string]struct{})
```

Comment:
Maybe use set.String?

Reply:
I have changed to `zoneSet := sets.Set[string]{}` :)